src/csharp/Grpc.Core/ServerCallContext.cs

```diff
@@ -29,45 +30,49 @@ namespace Grpc.Core
     /// </summary>
     public class ServerCallContext
     {
-        private readonly CallSafeHandle callHandle;
+        private readonly object extraData;
```

Comment:
Though this probably works but I'm not sure if it's the best way to separate the native types.

I assume this implementation is mostly concerned with back compat but ideally everything (types and methods) exposed publicly on `ServerCallContext` could have been defined on a `IServerCallContext` interface and a `NativeServerCallContext` would implement the interface with its own mechanism of storing the extra data and getters/setters for `WriteOptions`, `Peer`, `AuthContext`, etc.

Since we probably can't change all the generated code now to start accepting a `IServerCallContext` interface in the call handlers since that would be a major breaking change. I'm wondering if it's possible to make ServerCallContext an abstract class and move all the implementation into a `NativeServerCallContext` that extends from it?

I haven't tried this yet so I'm not sure about the ramifications but I'd like to explore that possibility before resorting to this mechanism with `objects` types and `Func` that you pass to the constructor to specify how to retrieve properties.

Though changing `ServerCallContext` from a POCO to an abstract base class is technically breaking, I think it's small enough of a breaking change to be worth it.

What do you think @shirhatti @jamesNK?

Reply:
You're right, this is all about backward compatibility, otherwise this change would be trivial.

Ad `IServerCallContext` interface:
- we definitely can't change `ServerCallContext` to `IServerCallContext` in the generated code. `ServerCallContext` is exposed in every server-side handler so a change would break a large portion of current users.
- using an interface for server call context is not right in general as there might be more members added to `ServerCallContext` over time (e.g. long time ago we added AuthContext) and adding new members to an interface is a breaking change (using ServerCallContext as a base class would be better as it could have default implementations for newly added members).

Ad changing `ServerCallContext` to an abstract class base:
- this is something we *could* consider, because currently there are no public constructors so the only instances of ServerCallContext in existence are the ones created in Grpc.Core or by Grpc.Core.Testing. I'd be worried about breaking the binary compatibility though (this has the potential to break all of the many existing nuget packages that depend on Grpc.Core).
See "adding virtual to a member" in https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#members
- I agree that just changing `ServerCallContext` to a base class is much simpler, but the solution I'm proposing is safer from the perspective of backward compatibility.

One possible solution would be to change ServerCallContext into a base class, leave the existing non-virtual members as is (adding "virtual" is breaking) and make them call a newly added protected virtual or abstract member (even abstract is fine as there are currently no accessible constructors outside of our control). These members can be then overridden by subclasses like 'NativeServerCallContext' that would provide the right behavior for each grpc implementation.
@apolcyn  @JunTaoLuo  WDYT?